### PR TITLE
fix(idempotency): POWERTOOLS_IDEMPOTENCY_DISABLED should respect truthy values

### LINKED ISSUE #4390: Bug: POWERTOOLS_IDEMPOTENCY_DISABLED does not respect truthy values
### Expected Behaviour

I would expect truthy values to be respected for `POWERTOOLS_IDEMPOTENCY_DISABLED`:

- True: ["1", "y", "yes", "t", "true", "on"]
- False: ["0": "n", "no", "f", "false", "off"]

### Current Behaviour

The logic below is accessing the environment variable and bypassing the idempotency layer if it the condition evaluates to `True`. But it's always evaluating to `True` unless `os.getenv` returns `None`.

```
if os.getenv(constants.IDEMPOTENCY_DISABLED_ENV):
    return function(*args, **kwargs)
```

### Code snippet

```python
>>> import os
>>> os.environ['POWERTOOLS_IDEMPOTENCY_DISABLED'] = 'false'
>>> bool(os.getenv('POWERTOOLS_IDEMPOTENCY_DISABLED'))
True
>>> bool(None)
False
```

### Possible Solution

Leverage the `strtobool` function in `aws_lambda_powertools.shared.functions` to convert the environment variable value into a `bool`.

### Steps to Reproduce

I would recommend setting the value of `POWERTOOLS_IDEMPOTENCY_DISABLED` to `false`, which implies the user does not want idempotency to be disabled. In this situation, the Idempotency layer will be bypassed and no records will be recorded in the idempotency store.

### Powertools for AWS Lambda (Python) version

latest

### AWS Lambda function runtime

3.11

### Packaging format used

Lambda Layers

### Debugging logs

_No response_

### PR DESCRIPTION
**Pull request by stevenhoelscher**
**Issue number:** [4390](https://github.com/aws-powertools/powertools-lambda-python/issues/4390)

### Changes

- When evaluating the `POWERTOOLS_IDEMPOTENCY_DISABLED` environment variable, first transform the string truthy value into a `bool`.

### User experience

* **Before**: A user sets `POWERTOOLS_IDEMPOTENCY_DISABLED=false` but the idempotency layer would actually be disabled
* **After**: A user sets `POWERTOOLS_IDEMPOTENCY_DISABLED=false` and the idempotency layer would not be disabled

## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [X] [Meet tenets criteria](https://docs.powertools.aws.dev/lambda/python/#tenets)
* [X] I have performed a self-review of this change
* [ ] Changes have been tested
* [ ] Changes are documented
* [X] PR title follows [conventional commit semantics](https://github.com/aws-powertools/powertools-lambda-python/blob/develop/.github/semantic.yml)

<details>
<summary>Is this a breaking change?</summary>

**RFC issue number**:

Checklist:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

</details>

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

### CODE REVIEW

**Comment by heitorlessa** [comment]
Looking into it shortly (in transit)

**Comment by leandrodamascena** [comment]
Hi @stevenhoelscher! Thank you so much for submitting this PR. It's a great catch, and I'm glad you discovered this long-standing bug. We really appreciate you taking the time to fix it!

I made only two changes in this PR:
1 - I set the default value of `os.getenv` to `false`, making the behavior more explicit.
2 - I included a test case to ensure the correct behavior.

Thanks

File: aws_lambda_powertools/utilities/idempotency/idempotency.py [hunk 1]
```diff
@@ -66,7 +67,9 @@ def idempotent(
         >>>     return {"StatusCode": 200}
     """
 
-    if os.getenv(constants.IDEMPOTENCY_DISABLED_ENV):
+    # Check if the IDEMPOTENCY_DISABLED_ENV environment variable is set to True
+    # If IDEMPOTENCY_DISABLED_ENV is not set or set to any false value, the function will be idempotent by default
```
Comment (heitorlessa) [inline aws_lambda_powertools/utilities/idempotency/idempotency.py]:
```suggestion
    # Skip idempotency controls when POWERTOOLS_IDEMPOTENCY_DISABLED has a truthy value
```

more concise

File: aws_lambda_powertools/utilities/idempotency/idempotency.py [hunk 2]
```diff
@@ -66,7 +67,9 @@ def idempotent(
...
+    if strtobool(os.getenv(constants.IDEMPOTENCY_DISABLED_ENV, "false")):
         return handler(event, context, **kwargs)
```
Comment (heitorlessa) [inline aws_lambda_powertools/utilities/idempotency/idempotency.py:81]: let's add a warning as it's not something customers would naturally want it disabled in production. It's easy to set and forget as we've learned from Metrics and other features.

File: aws_lambda_powertools/utilities/idempotency/idempotency.py [hunk 3]
```diff
@@ -150,7 +153,9 @@ def process_order(customer_id: str, order: dict, **kwargs):
 
     @functools.wraps(function)
     def decorate(*args, **kwargs):
-        if os.getenv(constants.IDEMPOTENCY_DISABLED_ENV):
+        # Check if the IDEMPOTENCY_DISABLED_ENV environment variable is set to True
+        # If IDEMPOTENCY_DISABLED_ENV is not set or set to any false value, the function will be idempotent by default
```
Comment (heitorlessa) [inline aws_lambda_powertools/utilities/idempotency/idempotency.py]:
```suggestion
        # Skip idempotency controls when POWERTOOLS_IDEMPOTENCY_DISABLED has a truthy value
```

File: aws_lambda_powertools/utilities/idempotency/idempotency.py [hunk 4]
```diff
@@ -150,7 +153,9 @@ def process_order(customer_id: str, order: dict, **kwargs):
...
+        if strtobool(os.getenv(constants.IDEMPOTENCY_DISABLED_ENV, "0")):
             return function(*args, **kwargs)
```
Comment (heitorlessa) [inline aws_lambda_powertools/utilities/idempotency/idempotency.py:173]: warning (as earlier)

**Review by heitorlessa — CHANGES_REQUESTED** [review]
let's add a warning to prevent customers forgetting it's disabled in dev but promoting it to prod.

File: aws_lambda_powertools/utilities/idempotency/idempotency.py (hunk 2 above)
Comment (leandrodamascena) [reply aws_lambda_powertools/utilities/idempotency/idempotency.py:81]: Hey @heitorlessa! I have created a new class, PowertoolsWarning, which inherits from UserWarning. This new class enables customers to filter and handle specific warnings raised by Powertools.

Comment (heitorlessa) [reply aws_lambda_powertools/utilities/idempotency/idempotency.py:81]: niiice! need to think of a good naming so the intent is clear. I'll come back today

File: aws_lambda_powertools/warnings/__init__.py [hunk 5]
```diff
@@ -0,0 +1,21 @@
+"""Shared warnings that don't belong to a single utility"""
+
+
+class PowertoolsWarning(UserWarning):
```
Comment (heitorlessa) [inline aws_lambda_powertools/warnings/__init__.py]:
```suggestion
class PowertoolsUserWarning(UserWarning):
```

This makes it more aligned with `UserWarning` as that's broadly applied to many types of warning. 

We should clean up the docstring to make it clearer and shorter -- what kind of use cases we raise this warning etc.

**Review by heitorlessa — CHANGES_REQUESTED** [review]

File: aws_lambda_powertools/warnings/__init__.py (hunk 5 above)
Comment (leandrodamascena) [reply aws_lambda_powertools/warnings/__init__.py]: Done.

Comment (leandrodamascena) [reply aws_lambda_powertools/warnings/__init__.py]: Done

**Comment by leandrodamascena** [comment]
@heitorlessa ready to review again.

File: aws_lambda_powertools/utilities/idempotency/idempotency.py [hunk 6]
```diff
@@ -66,7 +69,16 @@ def idempotent(
         >>>     return {"StatusCode": 200}
     """
 
-    if os.getenv(constants.IDEMPOTENCY_DISABLED_ENV):
+    # Skip idempotency controls when POWERTOOLS_IDEMPOTENCY_DISABLED has a truthy value
+    # Raises a warning if not running in development mode
+    if strtobool(os.getenv(constants.IDEMPOTENCY_DISABLED_ENV, "false")):
+        if not powertools_dev_is_set():
```
Comment (heitorlessa) [inline aws_lambda_powertools/utilities/idempotency/idempotency.py]: Think is necessary? It's another if branching here to maintain and higher memory

I'd personally warn either way. If they're explicitly in dev, other warnings will appear, and they can also suppress them.

File: aws_lambda_powertools/utilities/idempotency/idempotency.py [hunk 7]
```diff
@@ -150,7 +162,16 @@ def process_order(customer_id: str, order: dict, **kwargs):
 
     @functools.wraps(function)
     def decorate(*args, **kwargs):
-        if os.getenv(constants.IDEMPOTENCY_DISABLED_ENV):
+        # Skip idempotency controls when POWERTOOLS_IDEMPOTENCY_DISABLED has a truthy value
+        # Raises a warning if not running in development mode
+        if strtobool(os.getenv(constants.IDEMPOTENCY_DISABLED_ENV, "false")):
+            if not powertools_dev_is_set():
```
Comment (heitorlessa) [inline aws_lambda_powertools/utilities/idempotency/idempotency.py]: Same as before

File: aws_lambda_powertools/warnings/__init__.py [hunk 8]
```diff
@@ -0,0 +1,23 @@
+"""Shared warnings that don't belong to a single utility"""
+
+
+class PowertoolsUserWarning(UserWarning):
+    """
+    This class provides a custom Warning tailored for better clarity when certain situations occur.
```
Comment (heitorlessa) [inline aws_lambda_powertools/warnings/__init__.py:6]: That's a great docstring - straight to the point and actionable!

Well done 👏 

cc @sthulb who's learning the code style

**Review by heitorlessa — COMMENTED** [review]
Last question to clarify a piece I missed (changes are great!!)

File: aws_lambda_powertools/utilities/idempotency/idempotency.py (hunk 6 above)
Comment (leandrodamascena) [reply aws_lambda_powertools/utilities/idempotency/idempotency.py]: Agree. I'm sure we have situations like this in other parts of the codebase and I will review them.

**Review by heitorlessa — APPROVED** [review]

**Comment by heitorlessa** [comment]
hey @stevenhoelscher **thank you** so much for kicking this off <3

**Pull request merged**